partitura/musicanalysis/note_array_to_score.py

```diff
@@ -0,0 +1,358 @@
+            i, x in enumerate(note_array)]
+        return score.Score(partlist=parts)
+
+    if not isinstance(note_array, np.ndarray):
+        raise TypeError("The note array does not have the correct format.")
+
+    if len(note_array) == 0:
+        raise ValueError("The note array is empty.")
+
+    # Test Note array for negative durations
+    if "duration_div" in note_array.dtype.names:
+        assert np.all(note_array["duration_div"] >= 0), "Note array contains negative durations."
+    elif "duration_beat" in note_array.dtype.names:
+        assert np.all(note_array["duration_beat"] >= 0), "Note array contains negative durations."
+
+    # Note id creation or re-assignment
+    if "id" not in note_array.dtype.names:
+        note_ids = ["{}n{:4d}".format(name_id, i) for i in range(len(note_array))]
+        note_array = rfn.append_fields(note_array, "id", np.array(note_ids, dtype='<U256'))
+    elif assign_note_ids or np.all(note_array["id"] == note_array["id"][0]):
+        note_ids = ["{}n{:4d}".format(name_id, i) for i in range(len(note_array))]
+        note_array["id"] = np.array(note_ids)
+
+    dtypes = note_array.dtype.names
+    # check if note array contains time signatures
+    if not "ts_beats" in dtypes and key_sigs is None:
+        raise AttributeError("The note array does not contain a time signature.")
+
+    anacrusis_mask = np.zeros(len(note_array), dtype=bool)
+
+    # Start Checking Note array for available fields
+    if all([x in dtypes for x in ["onset_div", "pitch", "duration_div", "onset_beat", "duration_beat"]]):
+        anacrusis_mask[note_array["onset_beat"] < 0] = True
+    # This clause is related to bar normalized representations with global time stamps
+    elif all([x in dtypes for x in ["onset_beat", "duration_beat", "pitch", "global_score_time"]]):
+        anacrusis_mask[note_array["onset_beat"] < 0] = True
+        x = np.unique(note_array["onset_beat"])
+        renorm_onset_beat = False
+        if x.max() > note_array["ts_beats"].max():
+            pass
+        else:
+            for unique_ons in x:
+                tmp = note_array[note_array["onset_beat"] == unique_ons]
+                if not np.all(tmp["global_score_time"] == tmp[0]["global_score_time"]):
+                    renorm_onset_beat = True
+                    break
+
+        if renorm_onset_beat:
+            tmp = note_array[0]["onset_beat"]
+            tmp_ts = note_array[0]["ts_beats"]
+            for idx in range(1, len(note_array)):
+                if note_array[idx]["onset_beat"] < tmp:
+                    tmp_ts = tmp_ts + note_array[idx]["ts_beats"] if note_array[idx][
+                                                                         "onset_beat"] + tmp_ts < tmp else tmp_ts
+                    tmp = note_array[idx]["onset_beat"] + tmp_ts
+                    note_array[idx]["onset_beat"] = tmp
+        note_array = create_divs_from_beats(note_array)
```

Comment:
on a second look I don't understand this at all. what is the field "global_score_time"? is it used anywhere else in partitura? this also seems to be one of the only two places where the field `"ts_beats"` is required, the other being in connection with key sigs below, which I commented too. Maybe it's possible to get rid of this condition and estimate it, like in create_part?

Reply:
This was something that I implemented back when I started writing this, based on a couple of Basis features plus a normalization per bar, could also be useful with generation per bar models, and other staff. For now, I would say it is completely useless but I'm just including it for future use, in case we want to create a score from a generative algorithm.

Reply:
I vote against including that, as far as I can tell it's undocumented, untested, and unused.

Reply:
I will comment out this part of the code, however, this is something that would be very interesting to me and I would like to include it at some point.